File: 汇编语言与逆向工程/解码脚本/RC4.py
def rc4(ciphertext, key):
    """RC4 加解密（对称）"""
    # 初始化 S 盒
    sbox = list(range(256))
    j = 0
    key = [ord(c) if isinstance(c, str) else c for c in key]
    
    for i in range(256):
        j = (j + sbox[i] + key[i % len(key)]) % 256
        sbox[i], sbox[j] = sbox[j], sbox[i]
    
    # 生成密钥流并异或
    result = []
    i = j = 0
    for byte in ciphertext:
        i = (i + 1) % 256
        j = (j + sbox[i]) % 256
        sbox[i], sbox[j] = sbox[j], sbox[i]
        k = sbox[(sbox[i] + sbox[j]) % 256]
        result.append(byte ^ k)
    
    return bytes(result)

# 使用示例
ciphertext = bytes([0x3D, 0x93, 0x66, 0x29, 0xE7, 0x31, 0x0E, 0xB2,0x06,0x71,0x91,0x89,0x93])
key = "{This_is_a_key_!}"

plaintext = rc4(ciphertext, key)
print("Hex:", plaintext.hex().upper())
print("Bytes:", plaintext)

# 尝试解码为文本
try:
    print("Text:", plaintext.decode('utf-8'))
except:
    print("Text: (非 UTF-8 文本)")

File: 汇编语言与逆向工程/解码脚本/解码脚本/RC4.cpp
#include<stdio.h>
#include<windows.h>
#include<string.h>
#define MAX_KEY 256
#define MAX_STR 100

int pos_i=0;
int pos_j=0;
int sbox[256];
unsigned char key[MAX_KEY]=
"{This_is_a_key_!}";  // 密钥

void init_sbox();
unsigned char generate_key();
void decryption(unsigned char *ciphertext,int len,unsigned char *result);

int main()
{
 	SetConsoleOutputCP(65001); // 设置控制台输出为UTF-8编码
    SetConsoleCP(65001); // 设置控制台输入为UTF-8编码
	char result[MAX_STR]={0};
	int len;
	unsigned char cipher[MAX_STR]= {
0x3D,0x93,0x66,0x29,0xE7,0x31,0xF1,0x0E,0xB2,0x06,0x71,0x91,0x89,0x93,
    };  // 密文
	decryption((unsigned char*)cipher,14,(unsigned char*)result);//密文长度 
	printf("%s\n",result);
	return 0;
}

void decryption(unsigned char* ciphertext,int len,unsigned char*result) {
	pos_i =0;
	pos_j =0;
	int i=0;
	init_sbox();
	for(i=0;i<len;++i)
		result[i]= ciphertext[i]^(generate_key());
	result[i]='\0'; 
}



void init_sbox(){
    int i,j;
    int key_len = strlen((const char *)key);
    unsigned char temp;
    for(i=0;i<256;++i)
        sbox[i]=i;
    j=0;

    for(i=0;i<256;++i){
        j=(j+sbox[i]+key[i%key_len])%256;
        temp = sbox[i];
        sbox[i] = sbox[j];
        sbox[j] = temp;
    }
}

unsigned char generate_key(){
    unsigned char temp;
    int t;

    pos_i=(pos_i+1)%256;
    pos_j = (pos_j+sbox[pos_i])%256;
    temp = sbox[pos_i];
    sbox[pos_i] = sbox[pos_j];
    sbox[pos_j] = temp;
    t = (sbox[pos_i]+sbox[pos_j])%256;
    return sbox[t];
}
